### PR TITLE
[BCN] Add Type 2 fee's to api

### PR DESCRIPTION
**Pull request by leolambo**
Add type 2 fee support to BCN api.
Relies on this [crypto-rpc PR ](https://github.com/bitpay/crypto-rpc/pull/191) and a subsequent package update to support fetching priority fee.

### CODE REVIEW

**Review by kajoseph — CHANGES_REQUESTED** [review]
I would recommend adding a separate endpoint for priority fee estimation in `eth-routes.ts` and `matic-routes.ts`. That would clean up a lot of the branching logic for the EVM fee nuances. Since EVM now has 3 permutations for fee estimation, it gets somewhat messy trying to keep them all the same but separate. We can have the type0 gasPrice and type2 base fee use the same path, and separate priority fee.

Also, this seems to overlap with https://github.com/bitpay/bitcore/pull/3660. How do these reconcile?

File: packages/bitcore-node/src/providers/chain-state/evm/api/csp.ts [hunk 1]
```diff
@@ -106,16 +106,31 @@ export class BaseEVMStateProvider extends InternalStateProvider implements IChai
...
+    let { network, target = 4, txType, priorityFee } = params;
     const chain = this.chain;
     if (network === 'livenet') {
       network = 'mainnet';
     }
 
-    const cacheKey = `getFee-${chain}-${network}-${target}`;
+    let cacheKey = `getFee-${chain}-${network}`;
+    if (priorityFee) {
+      txType = txType || 2;
```
[the remark this answers is not in the document]
Comment (leolambo) [reply packages/bitcore-node/src/providers/chain-state/evm/api/csp.ts:117]: yes, priorityFee is not type 0 compatible

File: packages/bitcore-node/test/integration/ethereum/csp.spec.ts [hunk 2]
```diff
@@ -81,16 +81,43 @@ describe('Ethereum API', function() {
...
       target++;
     }
+    cacheKey = `getFee-${chain}-${network}-type2`;
+    const fee = await ETH.getFee({ chain, network, target, txType: 2});
+    expect(fee).to.exist;
+    const cached = await CacheStorage.getGlobal(cacheKey);
+    expect(fee).to.deep.eq(cached);
   });
 
+/*  it('should be able to get type 2 fees', async () => {
```
[the remark this answers is not in the document]
Comment (leolambo) [reply packages/bitcore-node/test/integration/ethereum/csp.spec.ts:100]: relied on this pr https://github.com/bitpay/crypto-rpc/pull/191. crypto rpc now needs to be updated

File: packages/bitcore-node/src/modules/matic/api/matic-routes.ts [hunk 3]
```diff
@@ -93,3 +93,20 @@ MaticRoutes.get('/api/MATIC/:network/ethmultisig/transactions/:multisigContractA
     return res.status(500).send(err);
   }
 });
+
+MaticRoutes.get('/api/ETH/:network/priorityFee/:percentile', async (req, res) => {
```
Comment (kajoseph) [inline packages/bitcore-node/src/modules/matic/api/matic-routes.ts:97]: MATIC

**Review by kajoseph — CHANGES_REQUESTED** [review]

File: packages/bitcore-node/src/providers/chain-state/evm/api/csp.ts [hunk 4]
```diff
@@ -106,16 +106,31 @@ export class BaseEVMStateProvider extends InternalStateProvider implements IChai
   }
 
   async getFee(params) {
-    let { network, target = 4 } = params;
+    let { network, target = 4, txType, priorityFeePercentile } = params;
```
Comment (kajoseph) [inline packages/bitcore-node/src/providers/chain-state/evm/api/csp.ts:109]: We should decouple the priority fee from this method altogether and make another method called like `getPriorityFee`. I don't think we need one mega function that handles all the things...having separate functions that are called in their own contexts would be ideal, I think

**Review by kajoseph — CHANGES_REQUESTED** [review]

**Review by kajoseph — APPROVED** [review]
tACK

**Pull request merged**